### PR TITLE
E2E Integration test environment setup

### LINKED ISSUE #639: Set up integration testing environment
## Problem
We need to setup an integration test environment for Mathesar, to test all user action scenarios.

## Proposed solution
Based on the discussions we've had previously (#89), the most suitable choice for us is [Playwright](https://playwright.dev/).

Integ tests require a complete mathesar setup, including a database. We should be able to test `user action -> api calls -> changes in db`.

## Additional context
- #89

### PR DESCRIPTION
**Pull request by pavish**
Fixes #639 

There are a few things to note:
* Playwright does not support linux arm64 images yet. It means that our integration tests will not work within dokcer in Mac M1 processors. Playwright team has mentioned that this will be fixed with their next major release.
* Playwright does not work with debian, mainly because the browser packages do not. Their official support is only for ubuntu. I tried modifying our docker setup to use ubuntu as the base, and install python, node and playwright over it. This lead to docker builds being super slow.
* I created a base image with ubuntu + python3.9 + node14 + playwright, and published it to dockerhub. I've used this image as our base image, so that our builds are faster.
* This image is currently in my account (did this to test it out) and with a free tier in dockerhub (only has limited pulls per day - around 200 for 6 hours): https://hub.docker.com/r/pavish73/mathesar-base/tags.
* I wanted to check if this approach sounds good to all of you. If it does, we can create an account for mathesar in dockerhub, and apply for their open source program: https://www.docker.com/community/open-source/application. This will give us unlimited pulls per day.
* The dockerfiles used to create the base imags can be placed in a dedicated repo.

## Checklist
 
- [ ] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [ ] My pull request targets the `master` branch of the repository
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by seancolsen — DISMISSED** [review]

**Comment by kgodey** [comment]
@pavish I accidentally pushed a GitHub workflow update to this branch. I think it's pretty harmless though, so I'll leave it there.

**Comment by pavish** [comment]
The failure in the pipeline is arising due to https://github.com/centerofci/mathesar/issues/787

**Comment by pavish** [comment]
@kgodey @mathemancer 

There are a couple things that we need to ensure as part of the image upload automation:
1. The image must get built and uploaded to dockerhub only if we detect changes on the dockerfile in the master branch
2. If there is a change on the dockerfiles, it needs to reflect on other workflows (pytest which depend on the dockerfile), on _all_ branches

I put some thought around this and came up with a few approaches.

**Approach 1**: I didn't want to spend a lot of time optimizing for this, so I took a step back from the image upload idea and tried just using a separate dockerfile with playwright directly for integ tests. Local docker build was so damn slow :/ It look 20 minutes for it to build for me. I suspect the github action for integ tests would slow down as well.

**Approach 2**: I realized we can have a step to upload images only on master branch, but for the integ test workflow we directly setup the environment on the workflow instead of relying on docker, as it is mentioned [here](https://playwright.dev/python/docs/ci#github-actions)
  - I haven't tested this out yet
  - This gives us another environment to maintain in addition to keeping the docker env and staging env in sync
 
**Approach 3**: I got this idea when going through the documentation for docker [build-push-action](https://github.com/docker/build-push-action/blob/master/docs/advanced/share-image-jobs.md).
  * Check if dockerfiles have changed. If yes: 
    - Build images, upload as artifact
    - If on master branch: Upload to dockerhub
  * While testing, pull from artifact
    - If present, load that docker image
    - If not present, it will automatically pull from dockerhub

I'm going to test both Approach 2 and 3. I do not want to take more time for this and if either of these approaches present more trouble, I'm going to settle with 1.

Let me know if you have any thoughts or suggestions around this.

**Comment by kgodey** [comment]
I think approach 3 sounds best but if it's going to take time to figure out, I think it's best to go with approach 1. We can always optimize later, the most important thing is to get the integration tests running.

**Comment by mathemancer** [comment]
Shouldn't we be checking for the docker cache breaking, rather than `Dockerfile` changes?  (or maybe I'm missing something).  However, whenever the cache _does_ break, I agree that Approach 3 sounds good long-run.  I also agree with @kgodey that just getting something working for integration tests (i.e., via Approach 1) should come first, and we can always get the more sophisticated setup working later.

File: .github/workflows/run-e2e-integ-tests.yml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      - name: Run type installation
+        run: docker exec mathesar_service python install.py --skip-confirm
+
+      - name: Run integ tests with pytest
+        run: docker exec mathesar_service pytest --browser chromium --browser webkit --browser firefox tests/
+
+      - uses: actions/upload-artifact@v2
+        if: ${{ failure() || success() }}
+        with:
+          name: Recordings
```
[the remark this answers is not in the document]
Comment (pavish) [reply .github/workflows/run-e2e-integ-tests.yml:55]: We may want to upload an artifact in a job and download in another job, time stamping the name would make it hard for us to do that.

I do understand the concern where multiple workflows could overwrite the artifact. So, I've modified the action to use different paths for each workflow in https://github.com/centerofci/mathesar/pull/746/commits/ba894a2d4a200cecfaef27e72fd24efa1367a702. Let me know if this looks good.

File: .github/workflows/run-pytest.yml [hunk 2]
```diff
@@ -33,4 +33,4 @@ jobs:
         run: docker-compose up --build -d
 
       - name: Run tests with pytest
-        run: docker exec mathesar_service pytest
+        run: docker exec mathesar_service pytest mathesar/
```
[the remark this answers is not in the document]
Comment (pavish) [reply .github/workflows/run-pytest.yml]: I've added it. Thanks.

**Comment by pavish** [comment]
@kgodey @mathemancer 

I have committed the Dockerfile directly for E2E tests.

This currently takes ~17 minutes to run on the workflow, and a lot more time locally. The other approaches require more thought. I also tried to simplify the docker setup by using binaries instead of compiling from source (for python and playwright) but that lead to different issues. I didn't want to go down that rabbit hole again and spend any more time on this PR.

File: .github/workflows/run-e2e-integ-tests.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (kgodey) [reply .github/workflows/run-e2e-integ-tests.yml:55]: Looks good.

File: .github/workflows/update-project-on-issue-label.yml [hunk 3]
```diff
@@ -43,6 +43,14 @@ jobs:
           cd .github/actions/project_update/
           python project_update.py ${{ github.event.issue.node_id }} --status Blocked
 
+      - name: Update Waiting issues
```
Comment (kgodey) [inline .github/workflows/update-project-on-issue-label.yml:46]: @pavish This change seems like a bad merge.

**Review by kgodey — APPROVED** [review]
@pavish Looks good. Please fix the issue with the issue label workflow and merge.

File: .github/workflows/update-project-on-issue-label.yml (hunk 3 above)
Comment (pavish) [reply .github/workflows/update-project-on-issue-label.yml:46]: @kgodey This was added in https://github.com/centerofci/mathesar/pull/746/commits/12ef48d436e33d47d60c33206b990205e47b38a5, which was mentioned here as harmless: https://github.com/centerofci/mathesar/pull/746#issuecomment-955142173

There wasn't another step with the same name. Should I go ahead and remove this?

Comment (kgodey) [reply .github/workflows/update-project-on-issue-label.yml:46]: @pavish it was superseded by https://github.com/centerofci/mathesar/commit/24e615a0c5f618d36e23ca508aa3a8b34036ac2b. I thought I had removed the older one. Please make sure that there's only one and it's the latest one (it should be setting the status to `Waiting` in the last line, not `Blocked`).

Comment (kgodey) [reply .github/workflows/update-project-on-issue-label.yml:46]: I think I got confused and forgot that I added it in this PR.

Comment (pavish) [reply .github/workflows/update-project-on-issue-label.yml]: Thanks. Updated in https://github.com/centerofci/mathesar/pull/746/commits/e5bfc70150b58c7b5a3854cb58cfab4ecaaac95d

**Pull request merged**